### ISSUE TITLE
Gradient checkpointing

### ISSUE THREAD

**Issue by cpehle** [norse/norse#39]
Currently we don't provide any support for checkpointing the forward integration, but instead save the whole forward dynamics for the backwards pass. This scales poorly with a growing sequence length. We should consider support for checkpointing the forward computation to limit memory requirements.

**Comment by Jegp** [comment]
I did a bit of digging and found out that gradient checkpointing requires the return type to be a `Variable`. And only tensors are compatible with variables. Meaning, we have three options:

1) Not use gradient checkpoints
2) Modify our states to return tuples of tensors instead of named tuples
3) Hack up our own version of the `checkpoint` function

I'm slightly torn here. On one hand I've had to struggle quite a bit with our named tuples, mostly in relation to sequential models. Pytorch just doesn't extend nicely in that direction, so it would be good to just go for 2) to increase compatibility. On the other, I really enjoy the named *and* typed tuples. Especially when the neurons become quite complex (the `LSNNState` has 4 entries, for instance).

Any inputs/opinions?

**Comment by xuniversus** [comment]
Is this still relevant/need it's own implementation? With the non-reentrant version it's possible to do checkpointing over time? At least it seems to work natively when I try it.

**Comment by xuniversus** [comment]
Not sure what the extension would be as many of the issues have been solved. Some interesting aspects of course is the number of timesteps as this drastically changes the memory footprint and dimension one need to consider to minimize it. If the goal of any extension were clear I could take a peek at it.
